_posts/2023-07-12-React-unity-BDD.markdown

```diff
@@ -0,0 +1,560 @@
+              curveDegrees: 0,
+            },
+            playerPosition: roundPlayerPosition(playerPosition),
+            delayTime: defaultDelayTime,
+            order: SceneIndex,
+          },
+        },
+        SceneIndex,
+      );
+    },
+    [SceneIndex, dispatch, siteCode, updatedAnimationDetails?.id],
+  );
+
+  // subscribes on receivedAnimationSceneDetails event and on receiving
+  // response with event name receivedAnimationSceneDetails from unity, it will update the app
+  // stage with new scene
+  useUnityEventEffect(
+    ({ playerPosition, screenshotData }) => {
+      if (waitingForUnity) {
+        const base64 = convertUnityImageToBase64(screenshotData);
+        onAddScene(roundPlayerPosition(playerPosition), base64);
+      }
+    },
+    [onAddScene, waitingForUnity],
+    "receivedAnimationSceneDetails",
+  );
+
+  const handleAddScene = () => {
+    setWaitingForUnity(true);
+    // React executes `get animation details` command on add scene
+    // and gets the scene back
+    getAnimationSceneDetails();
+  };
+
+  return (
+    <StyledAddSceneButton
+      id="add-scene-lock-button"
+      data-testid={`button-add-scene`}
+      variant="contained"
+      disableElevation
+      onClick={handleAddScene}
+      sx={{
+        justifyContent: "space-between",
+      }}>
+      <Box display="flex" marginLeft="-0.5em">
+        <AddRounded />
+        <Box marginLeft="0.2em">Add Scene</Box>
+      </Box>
+    </StyledAddSceneButton>
+  );
+}
+```
+
+![test results](../img/unity-react-bdd/bdd-test.png)
+
+By following the BDD approach and mocking the response from Unity's event handler, we can seamlessly develop our feature. Jest provides the capability to mock the Unity response, allowing us to simulate different scenarios during testing.
+
+In addition to this approach, there is another method to verify Unity commands. By mocking the Unity context provider, we can assert whether a specific event has been called with the correct event name. Here is an example of such a test:
+
+```tsx
+// jestMock.test.tsx
+import { executeCommand } from "../components/unityProvider/commandType";
+import UnityContextProvider from "../components/unityProvider/UnityContextProvider";
+import React, { FC } from "react";
+import { fireEvent, render, waitFor, screen } from "@testing-library/react";
+import { AnimationProvider } from "../components/animationProvider";
+import AnimationView from "../components/animationView";
+import { act } from "react-dom/test-utils";
+
+jest.mock("../components/unityProvider/commandType", () => ({
+  executeCommand: jest.fn(),
+}));
+
+const wrapper: FC<{}> = ({ children }) => (
+  <UnityContextProvider>
+    <AnimationProvider>{children}</AnimationProvider>
+  </UnityContextProvider>
+);
+
+describe("Should Call getAnimationSceneDetails command", () => {
+  it("should execute the getSceneDetails command when clicked add scene", async () => {
+    const executeCommandFn = jest.fn();
+    // @ts-ignore
```

Comment:
remove this comment, it's our old-tsconfig specificness which is why this is required, afresh project might not require this.